### PR TITLE
handle partial io

### PR DESCRIPTION
**Pull request by pmarks**
test_read_write_at will fail on linux without the retry logic, although you have to go up to an extremely large buffer. Presumably lustre will give you partial IO much more often.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -112,31 +112,70 @@ pub use range::Range;
...
-            *fd,
-            buf.as_mut_ptr() as *mut c_void,
-            buf.len() as size_t,
-            pos as off_t,
-        )
-    })
+    let mut total = 0usize;
+    let mut tries = 0usize;
+
+    while total < buf.len() && tries < 16 {
```
Comment (adam-azarchs) [inline src/lib.rs]: There's no good reason to limit the number of tries.

File: src/lib.rs [hunk 2]
```diff
@@ -112,31 +112,70 @@ pub use range::Range;
...
+                err(unsafe {
+                    pread(
+                        *fd,
+                        buf.as_mut_ptr().offset(total as isize) as *mut c_void,
+                        (buf.len() - total) as size_t,
+                        (pos + total as u64) as off_t,
+                )
+            }));
+
+        total += bytes;
```
Comment (adam-azarchs) [inline src/lib.rs]: You must first check that `bytes >= 0`.  If `bytes == 0` that's EOF, so you should break out.  If it's `< 0` then you've encountered an error.  But it's really quite critical that you not add bytes to total before checking that it's non-negative.

File: src/lib.rs [hunk 3]
```diff
@@ -112,31 +112,70 @@ pub use range::Range;
...
+        if bytes == 0 {
```
Comment (adam-azarchs) [inline src/lib.rs]: `bytes == 0` means EOF.  No reason to try again in that case.

**Comment by pmarks** [comment]
Ok gotcha.  The err() wrapper checks for negatives.

**Comment by adam-azarchs** [comment]
Ok, I see.  Sorry, still not used to rust.

If the err wrapper is checking for negative, then the only thing you should be checking for afterwards is 0, meaning EOF.   You should not limit the number of tries, since the OS is completely entitled to feeding you the bytes one at a time if it feels like it.

**Pull request merged**